src/configured_subscribed_notifications.c

```diff
@@ -0,0 +1,520 @@
+    char buf[26];
+    uint32_t i;
+
+    if (notif) {
+        /* find the top-level node */
+        while (notif->parent) {
+            notif = lyd_parent(notif);
+        }
+    }
+
+    switch (notif_type) {
+    case SR_EV_NOTIF_REPLAY_COMPLETE:
+        if (ATOMIC_INC_RELAXED(arg->replay_complete_count) + 1 < arg->sr_sub_count) {
+            /* wait until all the subscriptions finish their replay */
+            break;
+        }
+
+        /* WRITE LOCK on sub */
+        sub = sub_ntf_find_lock(arg->nc_sub_id, sub_id, 1);
+        if (!sub) {
+            EINT;
+            break;
+        }
+
+        /* context lock is held while the callback is executing */
+        ly_ctx = sr_acquire_context(np2srv.sr_conn);
+        sr_release_context(np2srv.sr_conn);
+
+        sprintf(buf, "%" PRIu32, sub->nc_sub_id);
+        lyd_new_path(NULL, ly_ctx, "/ietf-subscribed-notifications:replay-completed/id", buf, 0, &ly_ntf);
+        csn_send_notif(&arg->recv_info, arg->nc_sub_id, *timestamp, &ly_ntf, 1);
+
+        /* now send all the buffered notifications */
+        for (i = 0; i < arg->rt_notif_count; ++i) {
+            csn_send_notif(&arg->recv_info, arg->nc_sub_id, arg->rt_notifs[i].timestamp,
+                           &arg->rt_notifs[i].notif, 1);
+        }
+
+        /* UNLOCK */
+        sub_ntf_unlock(sub_id);
+
+        break;
+    case SR_EV_NOTIF_TERMINATED:
+        /* WRITE LOCK on sub */
+        sub = sub_ntf_find_lock(arg->nc_sub_id, sub_id, 1);
+        if (!sub) {
+            EINT;
+            break;
+        }
+
+        if (sub_ntf_del_sr_sub_is_last(sub, sub_id)) {
+            /* last SR subscription terminated, remove the whole NC subscription */
+            sub_ntf_terminate_sub(sub, NULL);
+        }
+
+        /* UNLOCK */
+        sub_ntf_unlock(sub_id);
+
+        /* subscription-terminated notif was already sent */
+        break;
+    case SR_EV_NOTIF_REALTIME:
+        if (ATOMIC_LOAD_RELAXED(arg->replay_complete_count) < arg->sr_sub_count) {
+            /* realtime notification received before replay has been completed, store in buffer */
+            np_ntf_add_dup(notif, timestamp, &arg->rt_notifs, &arg->rt_notif_count);
+        } else {
+            /* send the realtime notification */
+            csn_send_notif(&arg->recv_info, arg->nc_sub_id, *timestamp,
+                           (struct lyd_node **)&notif, 0);
+        }
+        break;
+    case SR_EV_NOTIF_REPLAY:
+        /* send the replayed notification */
+        csn_send_notif(&arg->recv_info, arg->nc_sub_id, *timestamp,
+                       (struct lyd_node **)&notif, 0);
+        break;
+    case SR_EV_NOTIF_MODIFIED:
+        /* handled elsewhere */
+        break;
+    case SR_EV_NOTIF_RESUMED:
+    case SR_EV_NOTIF_SUSPENDED:
+        /* should never be generated by the server */
+        EINT;
+        break;
+    }
+}
+
+static int
+csn_sr_subscribe(const char *stream, const char *xpath, const struct timespec *start,
```

Comment:
Seems a lot of the code was copied from the dynamic subscriptions, would be great if it is not duplicated.